atoms() method return value updated

mijo2 opened the pull request:
#### References to other Issues or PRs
Fixes #10152 and aims to complete the work started in 
PR #10246


#### Brief description of what is fixed or changed
--> Updated the atoms() method of class Basic to check if .args value of an object is empty or not to check if a sympy object is a leaf node or not.

#### Other comments
Let us consider an example to see how this PR has updated the .atoms() method. 
So, now we will define 3 basic sympy objects.

```
>>> b1 = Basic()
>>> b2 = Basic()
>>> b12 = Basic(b1, b2)
```

Now, before the fix, the output of ```b12.atoms()``` is:

```
>>> b12.atoms()
set()
```

After the fix, the output is:

```
>>> b12.atoms()
set({Basic()})
```
#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* core
  * Updated the definition of atoms method of class Basic
<!-- END RELEASE NOTES -->

mijo2 committed:
core: .atoms() method in Basic class now checks if .args is empty or not

Initially, .atoms() class roughly gave the list of objects that cannot
be further broken down as output but it didn't check if .args value of
the object is empty or not which is the correct way to check if something
is a leaf node or not.

mijo2 commented:
@asmeurer Some of the tests have to be updated for this functionality to be successful. 
For example - in ```test_cnodes.py``` the test:

```
def test_sizeof():
    typename = 'unsigned int'
    sz = sizeof(typename)
    assert ccode(sz) == 'sizeof(%s)' % typename
    assert sz.func(*sz.args) == sz
    assert not sz.is_Atom
    assert all(atom == typename for atom in sz.atoms())
```

Fails because earlier, ```sz.atoms() ``` returned an empty list but now it returns ```{String('unsigned int'), String('sizeof')}``` which are its arguments.

Should I proceed to update/remove tests like these?

oscarbenjamin commented:
I think that test should be changed perhaps to
```
assert sz.atoms() == ...
```

mijo2 committed:
codegen: Added an updated to an existing test case

A test case in test_cnodes.py used a test which considered
the old definition of .atoms() method. This test case was
updated accordingly.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -503,11 +503,11 @@ def atoms(self, *types):
-            if isinstance(expr, types):
+            if not (types or expr.args):
+                result.add(expr)
+            elif isinstance(expr, types):
```
I think it would be clearer to write:
```python
nodes = preorder_traversal(self)
if types:
    result = {node for node in nodes if isinstance(node, types)}
else:
    result = {node for node in nodes if not node.args}
```

mijo2 replied:
@oscarbenjamin sure thing.

mijo2 committed:
core: Minor code changes

A for loop was updated to make it more concise.

oscarbenjamin commented on sympy/codegen/tests/test_cnodes.py:
```diff
@@ -66,7 +66,7 @@ def test_sizeof():
     assert sz.func(*sz.args) == sz
     assert not sz.is_Atom
-    assert all(atom == typename for atom in sz.atoms())
+    assert sz.atoms() == set([String('unsigned int'), String('sizeof')])
```
You can create a set directly rather than creating a list and converting it to a set e.g. `{String(...), String(...)}`

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -503,12 +503,12 @@ def atoms(self, *types):
                 [t if isinstance(t, type) else type(t) for t in types])
-        else:
-            types = (Atom,)
         result = set()
```
I don't think this line is needed any more

mijo2 committed:
core: Some more changes

Some small changes were made in this commit as well.

oscarbenjamin commented:
Looks good
